### ISSUE TITLE
Should component state properties be enumerable?

### ISSUE THREAD

**Issue by patrick-steele-idem** [marko-js/marko#964]
I was looking into the following issue reported on StackOverflow: https://stackoverflow.com/questions/48035616/marko-js-on-submit

The developer is using the following code to submit an HTTP post:

```
  postSubmit() {
    axios.post('/api/v1/customers', this.state)
      .then(function (response) {
        console.log(response);
      })
    .catch(function (error) {
      console.log(error);
    });
  }
```

Because the properties on `this.state` are not enumerable, nothing ends up being included in the request.

I'm not seeing a good reason to not set `enumerable: true` when defining the properties: https://github.com/marko-js/marko/blob/b6f23b08868f1c3d3c15092cecf349ef9b898c7d/src/components/State.js#L6-L13

Thoughts?

**Comment by DylanPiercey** [comment]
Although i'm not totally against it I probably wouldn't personally recommend passing around state like this. For example if `axios` were to mutate the state object to add defaults for some of it's options this would re-render the component unnecessarily.

**Comment by bRRRITSCOLD** [comment]
Dylan how do you propose I pass around state then? Or should I just leave passing around state and solely create a new Object {} to pass around the data to the post request?

**Comment by DylanPiercey** [comment]
@bRRRITSCOLD If it were me in the case listed above I would create a new object to pass data to `axios`.

```js
  postSubmit() {
    axios.post('/api/v1/customers', {
        firstName: this.state.firstName
    })
      .then(function (response) {
        console.log(response);
      })
    .catch(function (error) {
      console.log(error);
    });
  }
```

I understand the convenience of directly passing state however and am curious what others think.

**Comment by bRRRITSCOLD** [comment]
I was thinking this too, but am new to marko and state and figured if it were there use it. But I completely back you idea of possible mutation and agree you are probably on to the right path.

**Comment by patrick-steele-idem** [comment]
You only need the `on-input` event handlers if you need to keep the component state synchronized with the input fields. This can be helpful you need to show validation errors as the user is typing. If you don't need that then you can just assign keys and use `getEl()` in `postSubmit()`:

```js
  postSubmit() {
    axios.post('/api/v1/customers', {
        firstName: this.getEl('firstName').value,
        // ...
    })
      .then(function (response) {
        console.log(response);
      })
    .catch(function (error) {
      console.log(error);
    });
  }
```

**Comment by bRRRITSCOLD** [comment]
And I would assume that this would circumvent the problem of passing around state that Dylan was speaking of?

**Comment by patrick-steele-idem** [comment]
Yes @bRRRITSCOLD, I think it is better to be explicit about what data is being sent to the backend. This allows you to put other things into `this.state` without worrying about it being submitted to the server.

**Comment by bRRRITSCOLD** [comment]
Have any good referral material or documentation on the theory of what should and should-be/should-not-be included in state? (What you should be explicit about and what you should not be worried about putting in state that will cause re-renders)

**Comment by patrick-steele-idem** [comment]
There's no "right" answer, but the Redux documentation has some good advice on organizing state: https://redux.js.org/docs/faq/OrganizingState.html

If you do decide to use Redux, I would recommend a new library: https://github.com/charlieduong94/marko-redux

Hope that helps!
